groups/nts/ntso/ntso_select.cpp

```diff
@@ -0,0 +1,639 @@
+
+    /// Unblock any thread waiting on the reactor when the specified
+    /// 'socket' is readable. Return the error.
+    ntsa::Error showReadable(ntsa::Handle socket) BSLS_KEYWORD_OVERRIDE;
+
+    /// Unblock any thread waiting on the reactor when the specified
+    /// 'socket' is writable. Return the error.
+    ntsa::Error showWritable(ntsa::Handle socket) BSLS_KEYWORD_OVERRIDE;
+
+    /// Do not unblock any thread waiting on the reactor when the specified
+    /// 'socket' is readable. Return the error.
+    ntsa::Error hideReadable(ntsa::Handle socket) BSLS_KEYWORD_OVERRIDE;
+
+    /// Do not unblock any thread waiting on the reactor when the specified
+    /// 'socket' is writable. Return the error.
+    ntsa::Error hideWritable(ntsa::Handle socket) BSLS_KEYWORD_OVERRIDE;
+
+    /// Block until at least one socket has a condition of interest or the
+    /// specified absolute 'deadline', if any, elapses. Load into the specified
+    /// 'result' the events that describe the sockets and their conditions.
+    /// Return the error.
+    ntsa::Error wait(
+        ntsa::EventSet*                                result,
+        const bdlb::NullableValue<bsls::TimeInterval>& deadline)
+        BSLS_KEYWORD_OVERRIDE;
+};
+
+void Select::copy(fd_set* destination, const fd_set& source)
+{
+#if defined(BSLS_PLATFORM_OS_UNIX)
+#if defined(FD_COPY)
+    FD_COPY(&source, destination);
+#else
+    *destination = source;
+#endif
+#elif defined(BSLS_PLATFORM_OS_WINDOWS)
+#if defined(FD_COPY)
+    FD_COPY(&source, destination);
+#else
+    destination->fd_count = source.fd_count;
+    bsl::memcpy(&destination->fd_array,
+                &source.fd_array,
+                source.fd_count * sizeof(source.fd_array[0]));
+#endif
+#else
+#error Not implemented
+#endif
+}
+
+Select::Select(const ntsa::ReactorConfig& configuration,
+               bslma::Allocator*          basicAllocator)
+: d_interestSet(basicAllocator)
+, d_maxHandle(0)
+, d_config(configuration, basicAllocator)
+, d_allocator_p(bslma::Default::allocator(basicAllocator))
+{
+    if (d_config.autoAttach().isNull()) {
+        d_config.setAutoAttach(false);
+    }
+
+    if (d_config.autoDetach().isNull()) {
+        d_config.setAutoDetach(false);
+    }
+
+    FD_ZERO(&d_readable);
+    FD_ZERO(&d_writable);
+    FD_ZERO(&d_exceptional);
+
+    NTSO_SELECT_LOG_DEVICE_CREATE();
+}
+
+Select::~Select()
+{
+    NTSO_SELECT_LOG_DEVICE_CLOSE();
+}
+
+ntsa::Error Select::attachSocket(ntsa::Handle socket)
+{
+    ntsa::Error error;
+
+    if (socket < 0) {
+        return ntsa::Error(ntsa::Error::e_INVALID);
+    }
+
+    error = d_interestSet.attach(socket);
+    if (error) {
+        return error;
+    }
+
+    FD_SET(socket, &d_exceptional);
+
+    if (socket > d_maxHandle) {
+        d_maxHandle = socket;
+    }
+
+    NTSO_SELECT_LOG_ADD(socket);
+
+    return ntsa::Error();
+}
+
+ntsa::Error Select::detachSocket(ntsa::Handle socket)
+{
+    ntsa::Error error;
+
+    if (socket < 0) {
```

Comment:
The same validation is done inside `d_interestSet.detach(socket);`

Reply:
We now defend against invalid handles only in ntsa::InterestSet.